language-plutus-core/exe/Main.hs

```diff
@@ -113,47 +179,115 @@ exampleSingle = ExampleSingle <$> exampleName
 exampleOpts :: Parser ExampleOptions
 exampleOpts = ExampleOptions <$> exampleMode
 
-printMode :: Parser PrintMode
-printMode = option auto
-  (  long "print-mode"
-  <> metavar "MODE"
-  <> value Classic
-  <> showDefault
-  <> help "Print mode: Classic -> plcPrettyClassicDef, Debug -> plcPrettyClassicDebug" )
 
-printOpts :: Parser PrintOptions
-printOpts = PrintOptions <$> input <*> printMode
+---------------- Reading programs from files ----------------
 
-runTypecheck :: TypecheckOptions -> IO ()
-runTypecheck (TypecheckOptions inp) = do
-    contents <- getInput inp
+-- Read a PLC source file
+getPlcInput :: Input -> IO String
+getPlcInput (FileInput file) = readFile file
+getPlcInput StdInput         = getContents
+
+-- Read and parse a PLC source file
+parsePlcFile :: Input -> IO ParsedProgram
+parsePlcFile inp = do
+    contents <- getPlcInput inp
     let bsContents = (BSL.fromStrict . encodeUtf8 . T.pack) contents
+    case PLC.runQuoteT $ runExceptT (PLC.parseScoped bsContents) of
+        Left (errCheck :: PlcParserError) -> do
+            T.putStrLn $ PLC.prettyPlcDefText errCheck
+            exitFailure
+        Right (Left (errEval :: PlcParserError)) -> do
+            print errEval
+            exitFailure
+        Right (Right (p :: ParsedProgram)) ->
+            return p
+
+-- Read a PLC AST from a CBOR file
+getCborInput :: Input -> IO BSL.ByteString
+getCborInput StdInput         = BSL.getContents
+getCborInput (FileInput file) = BSL.readFile file
+
+-- Read either a PLC file or a CBOR file, depending on 'fmt'
+getProg :: Input -> Format -> IO ParsedProgram
+getProg inp fmt =
+    case fmt of
+      Plc  -> parsePlcFile inp
+      Cbor -> do
+               p <- getCborInput inp
+               return $ fakeAlexPosn <$ (deserialise p :: PlainProgram)
+                   where fakeAlexPosn = PLC.AlexPn 0 0 0
```

Comment:
But we do serialize/deserialize annotations?

Reply:
How can we tell what type of annotations an AST has when we deserialise it?  The compiler complains if you don't annotate `deserialise` with the expected return type.

Reply:
Sure, but what is that type of annotations that we usually use? `()` like you did there? Or `AlexPn`? If it's the latter, then we can just call `deserialize` and call it a day. If it's `()`, then your code is fine. In any case we need a comment, preferably right in `plc --help` saying that any input is expected to contain annotations of some particular type.

Reply:
Good point.  Given the stuff I did on the sizes of scripts, I think we're assuming that the annotations will all be () for code that's transmitted to the chain, but once it gets there you can do what you want with the annotations.  In fact, in #1818 there's an issue to omit unit annotations from CBOR-serialised PLC altogether (and re-insert them automatically when we deserialise).  I don't think that ever got done, so I may take a look at it.  Doing that definitely saves a lot of space in serialised scripts, so I think the default should be to have unit annotations in PLC that we're going to serialise.

Reply:
OK, I've added a Note about this to the source and explained it in the help text as well.

Reply:
Yeah, I agree assuming unit seems reasonable. I guess it should be whatever our "standard" serialization format is, and while it's currently possible to serialize annotations, we don't normally do so since we haven't quite managed to preserve useful positions that far, and we definitely don't want to put them on chain.

(The issue about avoiding annotations is specifically about avoiding () annotations: if we have actual content in the annotations we should keep it!)